pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,308 @@
+	compareResultRangeSourceBefore
+	compareResultRangeSourceBeforeValue
+	compareResultRangeDestBefore
+	compareResultRangeDestBeforeValue
+	compareResultRangeIdentical
+	compareResultValueSourceBefore
+	compareResultValueDestBefore
+	compareResultValueIdentical
+	compareResultValueConflicting
+)
+
+func compareValues(sourceValue *graveler.ValueRecord, destValue *graveler.ValueRecord) compareResult {
+	c := bytes.Compare(sourceValue.Key, destValue.Key)
+	switch {
+	case c < 0:
+		return compareResultValueSourceBefore
+	case c > 0:
+		return compareResultValueDestBefore
+	case bytes.Equal(sourceValue.Identity, destValue.Identity):
+		return compareResultValueIdentical
+	default:
+		return compareResultValueConflicting
+	}
+}
+
+func compare(sourceValue *graveler.ValueRecord, sourceRange *Range, destValue *graveler.ValueRecord, destRange *Range) compareResult {
+	switch {
+	case sourceValue != nil && destValue != nil:
+		return compareValues(sourceValue, destValue)
+	case sourceValue != nil && bytes.Compare(sourceValue.Key, destRange.MinKey) < 0:
+		return compareResultValueSourceBefore
+	case sourceValue != nil && bytes.Compare(destRange.MaxKey, sourceValue.Key) < 0:
+		return compareResultRangeDestBefore
+	case sourceValue != nil:
+		return compareResultRangeDestBeforeValue
+	case destValue != nil && bytes.Compare(destValue.Key, sourceRange.MinKey) < 0:
+		return compareResultValueDestBefore
+	case destValue != nil && bytes.Compare(sourceRange.MaxKey, destValue.Key) < 0:
+		return compareResultRangeSourceBefore
+	case destValue != nil:
+		return compareResultRangeSourceBeforeValue
+	case bytes.Compare(sourceRange.MaxKey, destRange.MinKey) < 0:
+		return compareResultRangeSourceBefore
+	case bytes.Compare(destRange.MaxKey, sourceRange.MinKey) < 0:
+		return compareResultRangeDestBefore
+	case destRange.ID == sourceRange.ID:
+		return compareResultRangeIdentical
+	default:
+		return compareResultRangeConflict
+		// ranges are overlapping with different IDs
+	}
+}
+
+func proceedToRange(iter Iterator, key Key) error {
+	hasMore := true
+	for hasMore {
+		_, rng := iter.Value()
+		if rng == nil {
+			hasMore = iter.Next() // TODO(Guys): change
+		}
+		if rng == nil || bytes.Compare(key, rng.MaxKey) <= 0 {
+			return nil
+		}
+		hasMore = iter.NextRange()
+	}
+	return iter.Err()
+}
+
+func proceedToValue(iter Iterator, key Key) error {
+	hasMore := true
+	for hasMore {
+		record, rng := iter.Value()
+		if record == nil && rng == nil {
+			hasMore = iter.Next() // TODO(Guys): change
+			continue
+		}
+		switch {
+		case bytes.Compare(rng.MaxKey, key) <= 0:
+			hasMore = iter.NextRange()
+		case bytes.Compare(key, rng.MinKey) <= 0:
+			return nil
+		case record != nil && bytes.Compare(key, record.Key) <= 0:
+			return nil
+		default:
+			hasMore = iter.Next()
+		}
+
+	}
+	return iter.Err()
+}
+func getState(baseIter Iterator, sourceValue *graveler.ValueRecord, sourceRange *Range, destValue *graveler.ValueRecord, destRange *Range) (mergeState, error) {
```

Comment:
This function does change state, the name "get" is misleading.  Perhaps it should be called something like `compareAndMoveBase`.